### PR TITLE
[DX] Avoid reporting deprecated errors

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

File: src/Application/ApplicationFileProcessor.php [hunk 1]
```diff
@@ -113,4 +127,34 @@ private function printFile(File $file): void
...
     }
+
+    /**
+     * Inspired by @see https://github.com/phpstan/phpstan-src/blob/89af4e7db257750cdee5d4259ad312941b6b25e8/src/Analyser/Analyser.php#L134
+     */
+    private function configureCustomErrorHandler(): void
+    {
+        $errorHandlerCallback = function (int $code, string $message, string $file, int $line): bool {
+            $reportingLevel = error_reporting();
+            if ($reportingLevel === 0) {
```
Comment (samsonasik) [inline src/Application/ApplicationFileProcessor.php]: you should can use `(error_reporting() & $code) === 0)` or `(! (error_reporting() & $code))`, see

https://github.com/samsonasik/ErrorHeroModule/blob/18b691e880d6e068dd94d17bffcfdd54518d2a4a/src/HeroTrait.php#L112

I verify error handling on `ErrorHeroModule` by this:

```suggestion
            if (error_reporting() & $code === 0) {
```

File: src/Application/ApplicationFileProcessor.php [hunk 2]
```diff
@@ -113,4 +127,34 @@ private function printFile(File $file): void
...
     }
+
+    /**
+     * Inspired by @see https://github.com/phpstan/phpstan-src/blob/89af4e7db257750cdee5d4259ad312941b6b25e8/src/Analyser/Analyser.php#L134
+     */
+    private function configureCustomErrorHandler(): void
+    {
+        $errorHandlerCallback = function (int $code, string $message, string $file, int $line): bool {
+            error_reporting();
+            if ((error_reporting() & $code === 0) !== 0) {
```
Comment (samsonasik) [inline src/Application/ApplicationFileProcessor.php]: unneeded call and repetitive check with `0` can be removed.

```suggestion
            if ((error_reporting() & $code) === 0) {
```

**Review by samsonasik — APPROVED** [review]
I verified with check code:

```
@mkdir('test');
@mkdir('test');
```

and with the change : 

```php
if ((error_reporting() & $code) === 0) {
```

it run ok, only check `error_reporting()` will end up getting error:

```bash
 [ERROR] Could not process "/Users/samsonasik/www/rector-src/src/Application/FileProcessor.php" file, due to:           
         "mkdir(): File exists". On line: 26     
```

**Comment by samsonasik** [comment]
I created PR to PHPStan for it sync https://github.com/phpstan/phpstan-src/pull/889

**Comment by samsonasik** [comment]
Also referenced to PHP documentation https://www.php.net/manual/en/language.operators.errorcontrol.php#language.operators.errorcontrol

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Comment by samsonasik** [comment]
@TomasVotruba it possibly cause error on downgrade, ref https://github.com/rectorphp/rector-src/runs/4643749917?check_suite_focus=true#step:9:9039

```bash
 [ERROR] Could not process                                                      
         "phar:///home/runner/work/rector-src/rector-src/vendor/phpstan/phpstan/
         phpstan.phar/src/Type/Constant/ConstantStringType.php" file, due to:   
         "Uninitialized string offset -1". On line: 222                
```

Previously, the `Uninitialized string offset -1` is skipped, ref https://github.com/rectorphp/rector-src/runs/4643702959?check_suite_focus=true#step:9:5208

It seems `E_WARNING` needs to be added, I will check.

**Comment by samsonasik** [comment]
I created PR https://github.com/rectorphp/rector-src/pull/1579 to try fix on stopped by `Uninitialized string offset -1` on downgrade

**Pull request merged**